# Cleaning and Exploring Data

In this project, our aim is to predict the sale price of houses in King County as accurately as possible.
To find a accurate model for predictions, we need to clean and explore our data. In this notebook, we clean the data and find some results about our data set.

The findings about this notebook inserted at the end of the notebook.

## Importing necessary libraries

In [1]:
import numpy as np      # To use our np.arrays
import pandas as pd     # To use dataframes

# To plot
import matplotlib.pyplot as plt  
%matplotlib inline        
import seaborn as sns        


In [2]:
pd.options.display.max_columns=100 # To see the hidden columns in dataframe

In [3]:
df = pd.read_csv('kc_house_data.csv')  # To assign data to dataframe

# First Look to Dataframe

In [4]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      5650     1.0         NaN   0.0          3      7        1180   
1      7242     2.0         0.0   0.0          3      7        2170   
2     10000     1.0         0.0   0.0          3      6         770   
3      5000     1.0         0.0   0.0          5      7        1050   
4      8080     1.0         0.0   0.0          3      8        1680   

  sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0           0.0      1955           0.0    98178  47.5112 -122.257   
1         400.0      1951        1991.0    98125  47.7210 -122.319   
2           0.0      1933           NaN    98028  47.7379 -122.233   
3         910.0      1965           0.0    98136  47.5208 -122.393   
4           0.0      1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503

In [5]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.159700e+04  21597.000000  19221.000000  21534.000000  21597.000000   
mean   1.509941e+04      1.494096      0.007596      0.233863      3.409825   
std    4.141264e+04      0.539683      0.086825      0.765686      0.650546   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068500e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above      yr_built  yr_renovated       zipcode  \
count  21597.000000  21597.000000  21597.000000  17755.000000  21597.000000   
mean       7.657915   1788.596842   1970.999676     83.636778  98077.951845   
std        1.173200    827.759761     29.375234    399.946414     53.513072   
min        3.000000    370.000000   1900.000000      0.000000  98001.000000   
25%        7.000000   1190.000000   1951.000000      0.000000  98033.000000   
50%        7.000000   1560.000000   1975.000000      0.000000  98065.000000   
75%        8.000000   2210.000000   1997.000000      0.000000  98118.000000   
max       13.000000   9410.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000   21597.000000   21597.000000  
mean      47.560093   -122.213982    1986.620318   12758.283512  
std        0.138552      0.140724     685.230472   27274.441950  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471100   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.231000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

In [7]:
df.shape

(21597, 21)

# Focus on Columns 

In [8]:
df['id'].value_counts() # To check the same id`s for dublicated rows

795000620     3
1825069031    2
2019200220    2
7129304540    2
1781500435    2
             ..
7812801125    1
4364700875    1
3021059276    1
880000205     1
1777500160    1
Name: id, Length: 21420, dtype: int64

In [9]:
df.loc[df['id']==795000620]

id        date     price  bedrooms  bathrooms  sqft_living  \
17588  795000620   9/24/2014  115000.0         3        1.0         1080   
17589  795000620  12/15/2014  124000.0         3        1.0         1080   
17590  795000620   3/11/2015  157000.0         3        1.0         1080   

       sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
17588      6250     1.0         0.0   0.0          2      5        1080   
17589      6250     1.0         0.0   0.0          2      5        1080   
17590      6250     1.0         NaN   0.0          2      5        1080   

      sqft_basement  yr_built  yr_renovated  zipcode      lat    long  \
17588           0.0      1950           0.0    98168  47.5045 -122.33   
17589           0.0      1950           0.0    98168  47.5045 -122.33   
17590           0.0      1950           NaN    98168  47.5045 -122.33   

       sqft_living15  sqft_lot15  
17588           1070        6250  
17589           1070        6250  
17590           1070        6250

When we checked the same id numbers, we realized that some houses are sold two times in 2 years. This gives us idea about price changes so, we decided to keep these rows.

In [10]:
df['bedrooms'].value_counts()

3     9824
4     6882
2     2760
5     1601
6      272
1      196
7       38
8       13
9        6
10       3
11       1
33       1
Name: bedrooms, dtype: int64

In [12]:
df.loc[df['bedrooms']==33]

id       date     price  bedrooms  bathrooms  sqft_living  \
15856  2402100895  6/25/2014  640000.0        33       1.75         1620   

       sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
15856      6000     1.0         0.0   0.0          5      7        1040   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
15856         580.0      1947           0.0    98103  47.6878 -122.331   

       sqft_living15  sqft_lot15  
15856           1330        4700

We assume that 33 bedrooms in 1 floor house is not logical. So, we changed this with the most common value as 3. 
In 21597 row data, 1 row will not affect results. 

In [14]:
df['bedrooms'].replace(to_replace = 33, value =3 , inplace=True) # To change 33 to 3

# Checking Null Values and Changing Them

To plot and model our results, we need to get rid of null values. 

In [15]:
df.isna().sum() # To see total of nan values as column base

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

### 'waterfront' Column Clean up

In [17]:
df['waterfront'].value_counts() # To see most common data

0.0    19075
1.0      146
Name: waterfront, dtype: int64

We decided to fill Nan values with '0', because there are very less '1'.

In [18]:
df['waterfront'].fillna(value=0,inplace=True) #filling Nan with 0

### 'yr_renovated' Column Clean Up

In [19]:
df['yr_renovated'].value_counts()/df['yr_renovated'].count() # to see the %'s'

0.0       0.958096
2014.0    0.004112
2003.0    0.001746
2013.0    0.001746
2007.0    0.001690
            ...   
1946.0    0.000056
1959.0    0.000056
1971.0    0.000056
1951.0    0.000056
1954.0    0.000056
Name: yr_renovated, Length: 70, dtype: float64

Approximately, 96% of the house renovated years are zero. So, we filled our Nan's with zero.

In [20]:
df['yr_renovated'].fillna(value=0,inplace=True) # filling null values 

### 'view' Clean Up

63 rows in this data will not change the results. So, we dropped them. 

In [22]:
df.dropna(subset=['view'],inplace=True) #dropping 63 rows 

## Last Check for Null Values

In [23]:
df.isna().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

# Checking Numerical Values and Strings

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21534 entries, 0 to 21596
Data columns (total 21 columns):
id               21534 non-null int64
date             21534 non-null object
price            21534 non-null float64
bedrooms         21534 non-null int64
bathrooms        21534 non-null float64
sqft_living      21534 non-null int64
sqft_lot         21534 non-null int64
floors           21534 non-null float64
waterfront       21534 non-null float64
view             21534 non-null float64
condition        21534 non-null int64
grade            21534 non-null int64
sqft_above       21534 non-null int64
sqft_basement    21534 non-null object
yr_built         21534 non-null int64
yr_renovated     21534 non-null float64
zipcode          21534 non-null int64
lat              21534 non-null float64
long             21534 non-null float64
sqft_living15    21534 non-null int64
sqft_lot15       21534 non-null int64
dtypes: float64(8), int64(11), object(2)
memory usage: 3.6+ MB


## Changing 'sqft_basement' from string to float 

In [27]:
df['sqft_basement'].unique() 

We realized from unique values there are '?' for some columns. We also understood from other rows that sqft_living consists of sqft_basement and sqft_above. So, we substracted and find the values.

In [28]:
df['sqft_basement'].replace(to_replace = '?', value = df['sqft_living'] - df['sqft_above'], inplace=True)

## Checking Date Column

In [31]:
import datetime
df["yr_bought"] = pd.to_datetime(df["date"])

In [32]:
df['year'] = pd.DatetimeIndex(df['yr_bought']).year

In [33]:
df['year'].value_counts()

2014    14588
2015     6946
Name: year, dtype: int64

All the data belongs to two years. So, we decided the drop this column, because we think that it is not necessary. We have already known years.

In [51]:
df.drop(columns=['date'],inplace=True)


In [52]:
df.drop(columns=['yr_bought'],inplace=True)

In [53]:
df.drop(columns=['year'],inplace=True)

# First Explorations

- Our data belongs to just two years. But, these years are not very old. Today's prices are not far away from them. So, it can give us accurate price prediction models.
- There are some categorical data and binary values like 'waterfront'.
- There were Nan values or wrong values in our data and we changed them.
- Some columns consists of other two columns like 'sqft_living'
- There are some very old houses which built in 1900. Maybe, some area are very old and some of them renovated. Or, maybe the missing values filled with 1900. We do not know these assumptions so, we decided not to change them. Because, it can cause to lose data. 

Lastly, we write our data new, clean data set. After that, we will use this dataset.

In [56]:
df.to_csv('clean_data.csv')